### ISSUE TITLE
[answered] Question: IN_SEQUENCE() behavior with TIMES()?

### ISSUE THREAD

**Issue by rcdailey** [rollbear/trompeloeil#71]
If I use `TIMES()` with `IN_SEQUENCE()`, does the sequence only require the order to match for the first invocation of all functions in the sequence? For example:

```cpp
trompeloeil::sequence seq;
REQUIRE_CALL(mock, Foo1)
    .TIMES(AT_LEAST(1))
    .IN_SEQUENCE(seq);
REQUIRE_CALL(mock, Foo2)
    .IN_SEQUENCE(seq);
REQUIRE_CALL(mock, Foo3)
    .IN_SEQUENCE(seq);

// later...

mock.Foo1();
mock.Foo2();
mock.Foo1();
mock.Foo3();
```

Does this pass or fail? I think it might be worth it to add a note in the TIMES() section of the reference documentation note any special behavior with sequences. I'm not sure what to expect here.

**Comment by rollbear** [comment]
Using `.TIMES()` with `.IN_SEQUENCE()` is confusing at best, and especially when you have a (possibly open) interval for `.TIMES()`, as your example shows.

I'll try to improve the documentation.

To answer your question directly, you always see sequences as observed from a sequence object, and a sequence object can only move forward in what it allows. So, in your example above, a sequence violation is reported on the second call to `Foo1()`. It goes like this:

`mock.Foo1();`
this is the first call for the sequence object, so it is allowed. It says `AT_LEAST(1)`, so it may move to the next step, or it may repeat the same call.

`mock.Foo2();`
The current step in the sequence is `Foo1()`, but it is satisfied, so moving on to the next one is allowed. The next one is `Foo2()`, which matches this call, so everything is good.

`mock.Foo1();`
The current step in the sequence is `mock.Foo2()`. Is is satisfied and saturated, so the sequence object must move to the next step. The next step is `Foo3()`, which is a mismatch, so a sequence violation is reported.

**Comment by rcdailey** [comment]
Thank you. I think your response is fantastic. You don't necessarily have to use the example I proposed in my question, but I definitely feel like documenting this in terms of examples and an explanation break-down of each step and how it interacts with the sequence is the best way to go. Very valuable information. I hope you will make it available to a wider audience.

Maybe to prevent convoluting your documentation for these (what I consider) edge-cases, you might have an "Advanced" main section somewhere that handles more complex configurations like this (I haven't checked to see if you already have this, nor am I sure if you want to structure it this way). Anyway, just some thoughts.

Thanks again. Closing this issue.
